File: frontend/apps/base-app/types/customer.ts
// Customer domain types with TypeScript utilities
import { BaseEntitySnake, WithMetadata, WithCustomFields, WithTags, Address, ContactInfo, DateString, CustomerID } from './common';
import { PartialBy, RequiredBy } from './utils';

// Base customer type extending common patterns (using snake_case to match API)
export interface Customer extends BaseEntitySnake, WithMetadata, WithCustomFields, WithTags {
  id: CustomerID;
  customer_number: string;
  first_name: string;
  last_name: string;
  middle_name?: string;
  display_name?: string;
  company_name?: string;
  customer_type: CustomerType;
  tier: CustomerTier;
  status: CustomerStatus;

  // Contact info
  email: string;
  phone?: string;
  mobile?: string;
  website?: string;

  // Address fields (API uses snake_case)
  address_line_1?: string;
  address_line_2?: string;
  city?: string;
  state_province?: string;
  postal_code?: string;
  country?: string;

  // Business fields
  tax_id?: string;
  vat_number?: string;
  credit_limit?: number;
  payment_terms?: number;

  // Metrics
  lifetime_value: number;
  total_purchases: number;
  average_order_value: number;
  last_purchase_date?: DateString;
  first_purchase_date?: DateString;
  last_interaction?: DateString;

  // Relationships
  primary_contact_id?: string;
  account_manager_id?: string;
  parent_customer_id?: CustomerID;

  // Additional
  notes?: string;
  preferences?: CustomerPreferences;
  segments?: string[];
}

// Customer enums using const assertions for better type inference
export const CustomerTypes = {
  INDIVIDUAL: 'individual',
  BUSINESS: 'business',
  ENTERPRISE: 'enterprise',
  PARTNER: 'partner',
  VENDOR: 'vendor'
} as const;
export type CustomerType = typeof CustomerTypes[keyof typeof CustomerTypes];

export const CustomerTiers = {
  FREE: 'free',
  BASIC: 'basic',
  STANDARD: 'standard',
  PREMIUM: 'premium',
  ENTERPRISE: 'enterprise'
} as const;
export type CustomerTier = typeof CustomerTiers[keyof typeof CustomerTiers];

export const CustomerStatuses = {
  PROSPECT: 'prospect',
  ACTIVE: 'active',
  INACTIVE: 'inactive',
  SUSPENDED: 'suspended',
  CHURNED: 'churned',
  ARCHIVED: 'archived'
} as const;
export type CustomerStatus = typeof CustomerStatuses[keyof typeof CustomerStatuses];

// Customer preferences
export interface CustomerPreferences {
  language?: string;
  currency?: string;
  timezone?: string;
  communicationChannels?: CommunicationChannel[];
  marketingOptIn?: boolean;
  invoiceDelivery?: 'email' | 'mail' | 'both';
}

export type CommunicationChannel = 'email' | 'sms' | 'phone' | 'push' | 'mail';

// Using TypeScript utilities for create/update operations
export type CustomerCreateInput = RequiredBy<
  PartialBy<Customer,
    'id' | 'customer_number' | 'created_at' | 'updated_at' | 'lifetime_value' |
    'total_purchases' | 'average_order_value' | 'status' | 'tier' | 'customer_type'
  >,
  'first_name' | 'last_name' | 'email'
>;

export type CustomerUpdateInput = PartialBy<Customer,
  'id' | 'customer_number' | 'created_at' | 'updated_at'
>;

// Search and filter types
export interface CustomerSearchParams {
  query?: string;
  email?: string;
  status?: CustomerStatus | CustomerStatus[];
  customerType?: CustomerType | CustomerType[];
  tier?: CustomerTier | CustomerTier[];
  country?: string;
  city?: string;
  tags?: string[];
  segments?: string[];
  minLifetimeValue?: number;
  maxLifetimeValue?: number;
  hasRecentPurchase?: boolean;
  page?: number;
  pageSize?: number;
  sortBy?: CustomerSortField;
  sortOrder?: 'asc' | 'desc';
}

export type CustomerSortField =
  | 'displayName'
  | 'email'
  | 'createdAt'
  | 'lifetimeValue'
  | 'totalPurchases'
  | 'lastPurchaseDate';

// Customer metrics for dashboard (using snake_case to match API)
export interface CustomerMetrics {
  total_customers: number;
  active_customers: number;
  new_customers_this_month: number;
  churned_this_month?: number;
  average_lifetime_value: number;
  total_revenue: number;
  retention_rate?: number;
  churn_rate?: number;
  customers_by_tier?: Record<string, number>;
  customers_by_status?: Record<string, number>;
  customers_by_type?: Record<string, number>;
  growth_rate?: number;
  top_customers?: CustomerSummary[];
  top_segments?: Array<{ name: string; count: number }>;
}

export interface CustomerSummary {
  id: CustomerID;
  displayName: string;
  email: string;
  lifetimeValue: number;
  tier: CustomerTier;
  status: CustomerStatus;
}

// Customer activity and history
export interface CustomerActivity {
  id: string;
  customerId: CustomerID;
  type: ActivityType;
  description: string;
  timestamp: DateString;
  metadata?: Record<string, unknown>;
}

export type ActivityType =
  | 'purchase'
  | 'support_ticket'
  | 'email_sent'
  | 'email_opened'
  | 'login'
  | 'profile_update'
  | 'subscription_change'
  | 'note_added';

// Type guards
export function isBusinessCustomer(customer: Customer): boolean {
  return customer.customer_type === CustomerTypes.BUSINESS ||
         customer.customer_type === CustomerTypes.ENTERPRISE;
}

export function isActiveCustomer(customer: Customer): boolean {
  return customer.status === CustomerStatuses.ACTIVE;
}

export function isPremiumCustomer(customer: Customer): boolean {
  return customer.tier === CustomerTiers.PREMIUM ||
         customer.tier === CustomerTiers.ENTERPRISE;
}

File: src/dotmac/platform/graphql/types/customer.py
"""
GraphQL types for Customer Management.

Provides types for customers, activities, and notes with efficient
batched loading via DataLoaders to prevent N+1 query problems.
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any

import strawberry


@strawberry.enum
class CustomerStatusEnum(str, Enum):
    """Customer account status."""

    PROSPECT = "prospect"
    ACTIVE = "active"
    INACTIVE = "inactive"
    SUSPENDED = "suspended"
    CHURNED = "churned"
    ARCHIVED = "archived"


@strawberry.enum
class CustomerTypeEnum(str, Enum):
    """Type of customer account."""

    INDIVIDUAL = "individual"
    BUSINESS = "business"
    ENTERPRISE = "enterprise"
    PARTNER = "partner"
    VENDOR = "vendor"


@strawberry.enum
class CustomerTierEnum(str, Enum):
    """Customer tier/level for service differentiation."""

    FREE = "free"
    BASIC = "basic"
    STANDARD = "standard"
    PREMIUM = "premium"
    ENTERPRISE = "enterprise"


@strawberry.enum
class ActivityTypeEnum(str, Enum):
    """Types of customer activities."""

    CREATED = "created"
    UPDATED = "updated"
    STATUS_CHANGED = "status_changed"
    NOTE_ADDED = "note_added"
    TAG_ADDED = "tag_added"
    TAG_REMOVED = "tag_removed"
    CONTACT_MADE = "contact_made"
    PURCHASE = "purchase"
    SUPPORT_TICKET = "support_ticket"
    LOGIN = "login"
    EXPORT = "export"
    IMPORT = "import"


@strawberry.type
class CustomerActivity:
    """Customer activity/interaction record."""

    id: strawberry.ID
    customer_id: strawberry.ID
    activity_type: ActivityTypeEnum
    title: str
    description: str | None
    performed_by: strawberry.ID | None
    created_at: datetime

    @classmethod
    def from_model(cls, activity: Any) -> "CustomerActivity":
        """Convert SQLAlchemy model to GraphQL type."""
        return cls(
            id=strawberry.ID(str(activity.id)),
            customer_id=strawberry.ID(str(activity.customer_id)),
            activity_type=ActivityTypeEnum(activity.activity_type.value),
            title=activity.title,
            description=activity.description,
            performed_by=(
                strawberry.ID(str(activity.performed_by)) if activity.performed_by else None
            ),
            created_at=activity.created_at,
        )


@strawberry.type
class CustomerNote:
    """Note/comment about a customer."""

    id: strawberry.ID
    customer_id: strawberry.ID
    subject: str
    content: str
    is_internal: bool
    created_by_id: strawberry.ID | None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_model(cls, note: Any) -> "CustomerNote":
        """Convert SQLAlchemy model to GraphQL type."""
        return cls(
            id=strawberry.ID(str(note.id)),
            customer_id=strawberry.ID(str(note.customer_id)),
            subject=note.subject,
            content=note.content,
            is_internal=note.is_internal,
            created_by_id=strawberry.ID(str(note.created_by_id)) if note.created_by_id else None,
            created_at=note.created_at,
            updated_at=note.updated_at,
        )


@strawberry.type
class CustomerMetrics:
    """Aggregated customer metrics."""

    lifetime_value: Decimal
    total_purchases: int
    average_order_value: Decimal
    last_purchase_date: datetime | None
    first_purchase_date: datetime | None
    satisfaction_score: int | None
    net_promoter_score: int | None


@strawberry.type
class Customer:
    """
    Customer with comprehensive profile and related data.

    Activities and notes are batched via DataLoaders to prevent N+1 queries.
    """

    # Core identifiers
    id: strawberry.ID
    customer_number: str

    # Basic information
    first_name: str
    last_name: str
    middle_name: str | None
    display_name: str | None
    company_name: str | None

    # Account information
    status: CustomerStatusEnum
    customer_type: CustomerTypeEnum
    tier: CustomerTierEnum

    # Contact information
    email: str
    email_verified: bool
    phone: str | None
    phone_verified: bool
    mobile: str | None

    # Address
    address_line1: str | None
    address_line2: str | None
    city: str | None
    state_province: str | None
    postal_code: str | None
    country: str | None

    # Business info (for business customers)
    tax_id: str | None
    industry: str | None
    employee_count: int | None

    # Metrics
    lifetime_value: Decimal
    total_purchases: int
    average_order_value: Decimal
    last_purchase_date: datetime | None

    # Dates
    created_at: datetime
    updated_at: datetime
    acquisition_date: datetime
    last_contact_date: datetime | None

    # Relationships - batched via DataLoaders
    activities: list[CustomerActivity] = strawberry.field(default_factory=list)
    notes: list[CustomerNote] = strawberry.field(default_factory=list)

    @classmethod
    def from_model(cls, customer: Any) -> "Customer":
        """Convert SQLAlchemy Customer model to GraphQL type."""
        return cls(
            id=strawberry.ID(str(customer.id)),
            customer_number=customer.customer_number,
            first_name=customer.first_name,
            last_name=customer.last_name,
            middle_name=customer.middle_name,
            display_name=customer.display_name,
            company_name=customer.company_name,
            status=CustomerStatusEnum(customer.status.value),
            customer_type=CustomerTypeEnum(customer.customer_type.value),
            tier=CustomerTierEnum(customer.tier.value),
            email=customer.email,
            email_verified=customer.email_verified,
            phone=customer.phone,
            phone_verified=customer.phone_verified,
            mobile=customer.mobile,
            address_line1=customer.address_line1,
            address_line2=customer.address_line2,
            city=customer.city,
            state_province=customer.state_province,
            postal_code=customer.postal_code,
            country=customer.country,
            tax_id=customer.tax_id,
            industry=customer.industry,
            employee_count=customer.employee_count,
            lifetime_value=customer.lifetime_value,
            total_purchases=customer.total_purchases,
            average_order_value=customer.average_order_value,
            last_purchase_date=customer.last_purchase_date,
            created_at=customer.created_at,
            updated_at=customer.updated_at,
            acquisition_date=customer.acquisition_date,
            last_contact_date=customer.last_contact_date,
            # activities and notes populated by resolvers via DataLoaders
            activities=[],
            notes=[],
        )


@strawberry.type
class CustomerConnection:
    """Paginated customer results."""

    customers: list[Customer]
    total_count: int
    has_next_page: bool


@strawberry.type
class CustomerOverviewMetrics:
    """High-level customer metrics."""

    total_customers: int
    active_customers: int
    prospect_customers: int
    churned_customers: int
    total_lifetime_value: Decimal
    average_lifetime_value: Decimal
